### ISSUE TITLE
pdf output misses unicode and combined inside code cells

### ISSUE THREAD

**Issue by Ken-B** [jupyter/nbconvert#533]
I develop in Julia which allows unicode and combined characters in code. However, they are missing when converting to pdf (html output works fine).

To replicate, create a new (Julia) notebook and inside a *code cell* type (using tab completion or just copy) for example
```julia
a = 1
â = 2
β = 3
β̂ = 4
```
(it seems even github has problems with the last one).

The pdf outputs:
a=1
a=2 
=3 
=4

So the hat is removed on the second line and the characters disappear completely in the last two lines.

As a workaround I first convert to html and then print to pdf, but then all the beautiful pdf layout is missing. I'm on nbconvert 5.1.1 (anaconda distribution) on mac osx 10.12.3. Let me know how I can help, thanks.

**Comment by takluyver** [comment]
I wonder if the font used in Latex for code doesn't have symbols for those characters.

**Comment by michaelpacer** [comment]
My guess is that it's a font issue, but there are a couple of other possibilities that I want to think through first: 

How LaTeX  handles combining characters like [	̂ ](http://www.fileformat.info/info/unicode/char/0302/index.htm) in math mode (at least traditionally) is using Unicode, and while this isn't a mathmode problem, I want to make sure that a couple of things work. Do you run into this issue when you use LaTeX in a markdown cell and directly invoke `$\hat{\beta}$`? Do your characters work when placed into a plain markdown cell? Do they work if you put the Unicode characters directly into a mathmode (e.g., `$β̂ = 4$`)? Do they work if you put them inside a literal `Verbatim` block?  

Depending on how all that goes we will have an idea of how to move forward. 

Also, the Latin characters that are displaying successfully are in the same Unicode plane so it shouldn't be _that_ kind of Unicode problem. I think instead it might be that the monospace font we're using in Verbatim doesn't include these Unicode characters.

**Comment by Ken-B** [comment]
> when you use LaTeX in a markdown cell and directly invoke $\hat{\beta}$

works, outputs characters

> into a plain markdown cell

doesn't work, same behaviour as in code cell, characters dropped

> Unicode characters directly into a mathmode (e.g., $β̂ = 4$)

doesn't work, same behaviour as in code cell, characters dropped

> inside a literal Verbatim block

doesn't work, same behaviour as in code cell, characters dropped

Are able to reproduce yourself or do you think it's a local issue on my end?
I'm more than happy to assist further.

**Comment by michaelpacer** [comment]
Ok, so this looks like its not specific to the Julia notebook either. I'm able to reproduce in a python notebook, just with markdown cells. I've spent a little time looking into it and I know that if I use a class that sets the roman font to one that I know has decent unicode support (minion pro) I get (mostly) expected behaviour back.

I will mention that even with Adobe's minion pro, it's not combining the `β` and the` ̂` correctly, and that is a font with a lot of features, so we may not be able to accommodate this easily. 

That said, we should _definitely_ have support for non-combining unicode characters, so I'm going to spend time figuring this out.

**Comment by michaelpacer** [comment]
Ok, so I've partially traced this down to the use of the `mathpazo` package without explicitly invoking Palatino (or any particular font) as the mainfont (i.e., via   `\setmainfont{}`). 

However, it still doesn't address the unicode inside math issue, which I know I fixed for my dissertation (and not in a completely hacky way). But I'll make a PR with the first bit of a fix. 

Question: @takluyver I'm not able to find anywhere the list of fonts that are going to be installed with  a default TeXLive install…I had looked this up for a few hours a couple weeks back too and I'm running into the same dead ends. Should we look into including a font as part of nbconvert itself?

**Comment by michaelpacer** [comment]
Some further updates: we need to include fontspec to get the combining character support (in general). Using the sourcecodepro package allows us to get these characters supported, but does not correctly position the circumflex above the β (much like in many places online, such as GitHub (β̂)). 

However, I fear weirdness coming from using so many different font declaration mechanisms and would prefer to just use fontspec for everything if we're going to be using it at all.

I made an [post on the  tex stackexchange](http://tex.stackexchange.com/questions/355081/fonts-using-xetex-why-do-fontspec-and-setmonofont-versus-sourcecodepro-differ), I'll keep you updated as I learn more.

**Comment by takluyver** [comment]
> Question: @takluyver I'm not able to find anywhere the list of fonts that are going to be installed with a default TeXLive install…I had looked this up for a few hours a couple weeks back too and I'm running into the same dead ends. Should we look into including a font as part of nbconvert itself?

I may answer this question with several of my own, as I don't know much about Latex.

What kind of font files does Latex use? Are they the same across different platforms and different Tex distributions? How big would one be that includes all of the characters people are likely to want (which is probably all the characters)? What's required to tell it about a font outside of the normal fonts location? Is *this* the same across different platforms and Tex distributions?

If we really can't rely on a sufficiently complete default font being available to Latex, then I think it makes sense to look at including one. But I'm somewhat reluctant - if we start shipping anything that a Latex installation might be missing, then we risk become a Tex distribution ourselves.

**Comment by michaelpacer** [comment]
I don't think that there is a font in existence that contains "all the characters".

Because we're using XeLaTeX it should be able to find any fonts in system locations (this can actually cause problems in some cases, see that [tex.stackexchange post](http://tex.stackexchange.com/questions/355081/fonts-using-xetex-why-do-fontspec-and-setmonofont-versus-sourcecodepro-differ) for an example with regards to Source Code Pro).

**Comment by takluyver** [comment]
> Thus if we could install a font as part of nbconvert to the system location, we should be able to always find it.

Could be tricky. If nbconvert is installed into an environment, it can't modify anything outside that environment. On Linux, we might be able to set `$XDG_DATA_DIRS` to affect where it looks for fonts (if xelatex uses that to find them). I don't know if there's an equivalent on other platforms. We could install it when nbconvert is run, but randomly installing stuff which affects the rest of the system is generally a bad idea.

**Comment by msto** [comment]
Hi,

I'm encountering this issue as well. (Jupyter v4.4.0, nbconvert v5.3.1). Is there a suggested fix or workaround?

Thanks!

**Comment by kmundnic** [comment]
One workaround is to use XeLatex and set the mono font for one that supports unicode, as suggested here: https://tex.stackexchange.com/questions/264390/xelatex-unicode-symbols-do-not-show-up-in-verbatim

**Comment by ibehnam** [comment]
This issue still exists! Any work on that by the devs?

**Comment by parsiad** [comment]
I can confirm that the workaround suggested by @kmundnic fixes the issue. Adding a bit more detail to their answer, you will have to do the following:
1. Create a [custom template](https://nbconvert.readthedocs.io/en/latest/customizing.html) and that includes `\setmonofont{DejaVu Sans Mono}`. The easiest way to do this is to copy an existing template and edit it.
2. Compile with `jupyter nbconvert --template <YOUR_TEMPLATE> ...`

**Comment by djsmedes** [comment]
The issue still exists. I got around it using a custom template as suggested, but it was much harder than I thought to implement -- mainly because nbconvert seems to use some kind of non-standard delimiters for the Jinja2 templates intended to be used with LaTeX (so as not to conflict with LaTeX's own special control character sequences, presumably). I don't know much about Jinja2, and the nbconvert docs don't even remotely hint that you'd be able to or want to do something like this.

Anyway, I eventually figured it out and put the whole thing into a Docker container. I made it a public repo so anyone can hopefully use it and it will "just work" and you won't have to go through the hours of troubleshooting I did.

https://github.com/djsmedes/julia-nbconvert-helper
